backend/app/assets/javascripts/spree/backend/line_items.js.coffee

```diff
@@ -1,55 +1,98 @@
-  editingDone(e)
-
-$(document).ready ->
-  $('.line-item')
-    .on('click', '.edit-line-item',   editing)
-    .on('click', '.cancel-line-item', editingDone)
-    .on('click', '.save-line-item',   onSaveLineItem)
-    .on('click', '.delete-line-item', onDeleteLineItem)
-
-lineItemURL = (id) ->
-  "#{Spree.routes.line_items_api(order_number)}/#{id}.json"
-
-adjustLineItem = (line_item_id, quantity) ->
-  url = lineItemURL(line_item_id)
-  Spree.ajax(
-    type: "PUT",
-    url: url,
-    data:
-      line_item:
-        quantity: quantity
-  ).done (msg) ->
-    window.Spree.advanceOrder()
-
-deleteLineItem = (line_item_id) ->
-  url = lineItemURL(line_item_id)
-  Spree.ajax(
-    type: "DELETE"
-    url: url
-  ).done (msg) ->
-    $('#line-item-' + line_item_id).remove()
-    if $('.line-items tr.line-item').length == 0
-      $('.line-items').remove()
-    window.Spree.advanceOrder()
+
+Spree.CartLineItemView = Backbone.View.extend
+  tagName: 'tr'
+  className: 'line-item'
+
+  initialize: (options) ->
+    @editing = options.editing || @model.isNew()
+    @noCancel = options.noCancel
+
+  events:
+    'click .edit-line-item': 'onEdit'
+    'click .cancel-line-item': 'onCancel'
+    'click .save-line-item': 'onSave'
+    'click .delete-line-item': 'onDelete'
+    'change .js-select-variant': 'onChangeVariant'
+
+  onEdit: (e) ->
+    e.preventDefault()
+    @editing = true
+    @render()
+
+  onCancel: (e) ->
+    e.preventDefault()
+    @trigger('cancel')
+    if @model.isNew()
+      @remove()
+    else
+      @editing = false
+      @render()
+
+  validate: () ->
+    @$('[name=quantity]').toggleClass 'error', !@$('[name=quantity]').val()
+    @$('.select2-container').toggleClass 'error', !@$('[name=variant_id]').val()
+
+    !@$('.select2-container').hasClass('error') && !@$('[name=quantity]').hasClass('error')
+
+  onSave: (e) ->
+    e.preventDefault()
+    return unless @validate()
+    attrs = {
+      quantity: parseInt(@$('input.line_item_quantity').val())
+    }
+    if @model.isNew()
+      attrs['variant_id'] = @$("[name=variant_id]").val()
+    @model.save attrs,
+      patch: true,
+      success: =>
+        window.Spree.advanceOrder()
+    @editing = false
+    @render()
+
+  onDelete: (e) ->
+    e.preventDefault()
+    return unless confirm(Spree.translations.are_you_sure_delete)
+    @remove()
+    @model.destroy
+      success: =>
+        window.Spree.advanceOrder()
+
+  render: ->
+    line_item = @model.attributes
+    image = line_item.variant && line_item.variant.images[0]
+    html = HandlebarsTemplates['orders/line_item'](
+      line_item: line_item,
+      image: image,
+      editing: @editing,
+      isNew: @model.isNew(),
+      noCancel: @noCancel
+    )
+    el = @$el.html(html)
+    @$("[name=variant_id]").variantAutocomplete({ in_stock_only: true })
+
+$ ->
+  if $("table.line-items").length
+    url = Spree.routes.orders_api + "/" + order_number
+    lineItemModel = Backbone.Model.extend
```

Comment:
Can we put this model into its own file and use the Spree.Model namespace discussed in #1715?